### PR TITLE
Remove SourcePawn and replace it with JavaScript

### PR DESCRIPTION
**Pull request by dvander**
As we all know, and many people have pointed out time and time again, SourcePawn is old and useless. Newer languages like JavaScript are vastly superior. Let's just remove SourcePawn entirely and replace it with angryzor's v8 branch. We should backport this to 1.7 since SourcePawn won't be needed there either.

### CODE REVIEW

**Comment by WildCard65** [comment]
Hahaha lol

**Comment by 50Wliu** [comment]
:+1: :star2: :stuck_out_tongue:

**Comment by Freeman-AM** [comment]
It's seems to be nice :ship:
:fish:

**Comment by chauffer** [comment]
:shipit:

**Comment by InstantMuffin** [comment]
Hi chauffer

File: AMBuildScript [hunk 1]
```diff
@@ -497,6 +503,7 @@ BuildScripts = [
   'extensions/updater/AMBuilder',
   'sourcepawn/compiler/AMBuilder',
   'sourcepawn/jit/AMBuilder',
```
Comment (asherkin) [inline AMBuildScript]: Dead code, don't need to build the SP VM or compiler any more.

File: core/logic/DebugReporter.h [hunk 2]
```diff
@@ -37,7 +37,7 @@
 
 class DebugReport : 
 	public SMGlobalClass, 
-	public IDebugListener
+	public IDebugListenerV8
```
Comment (asherkin) [inline core/logic/DebugReporter.h]: This seems messy, can just change IDebugListener directly.

File: core/logic/PluginSys.cpp [hunk 3]
```diff
@@ -42,6 +42,7 @@
 #include "GameConfigs.h"
 #include "common_logic.h"
 #include "Translator.h"
+#include <string>
```
Comment (asherkin) [inline core/logic/PluginSys.cpp]: Policy disallows STL in shipping SM code.

File: core/logic/PluginSys.cpp [hunk 4]
```diff
@@ -888,6 +889,16 @@ void CPluginManager::LoadPluginsFromDir(const char *basedir, const char *localpa
 	libsys->CloseDirectory(dir);
 }
 
+// TODO: Make better. Reason for not a method: would alter CPluginManager interface, which some extensions depend on
+static bool IsV8Plugin(CPlugin* plugin)
```
Comment (asherkin) [inline core/logic/PluginSys.cpp]: All plugins will be "V8" plugins.

File: core/logic/PluginSys.cpp [hunk 5]
```diff
@@ -921,43 +932,61 @@ LoadRes CPluginManager::_LoadPlugin(CPlugin **aResult, const char *path, bool de
...
 		{
-			if (error)
+			if(err != NULL)
 			{
-				smcore.Format(error, 
-					maxlength, 
-					"Unable to load plugin (error %d: %s)", 
-					err, 
-					g_pSourcePawn2->GetErrorString(err));
+				smcore.LogError("Could not load V8 plugin due to errors: %s",err);
```
Comment (asherkin) [inline core/logic/PluginSys.cpp]: Style: should be a space after the comma.

**Comment by asherkin** [comment]
Some comments in-line, this needs to be rebased before it can be merged however.
Should probably also at minimum have an implementation of admin-flatfile in JS before making this switch.

Please keep PRs to technical discussion or at least directly relating to the change being made.

**Comment by KyleSanderson** [comment]
I believe Matheus28 has already ported some of these plugins, his repository was called SourceMod.js.

**Comment by VoiDeD** [comment]
When can we expect to see ES6 features?

**Comment by psychonic** [comment]
:japanese_goblin:

**Comment by dvander** [comment]
sorry guys the Merge button is greyed out so I guess we can't do this.

**Comment by chauffer** [comment]
Oh. :crying_cat_face:

**Pull request closed without merging**